app/src/test/java/org/simple/clinic/bp/entry/BloodPressureValidationMockDateValidatorTest.kt

```diff
@@ -0,0 +1,168 @@
+  private val dateValidator = mock<UserInputDateValidator>()
+  private val bpValidator = BpValidator()
+
+  private val uiEvents = PublishSubject.create<UiEvent>()
+  private val patientUuid = UUID.fromString("79145baf-7a5c-4442-ab30-2da564a32944")
+
+  private val testUserClock = TestUserClock()
+  private val testUtcClock = TestUtcClock()
+  private val userSession = mock<UserSession>()
+
+  private val facilityRepository = mock<FacilityRepository>()
+  private val user = PatientMocker.loggedInUser(uuid = UUID.fromString("1367a583-12b1-48c6-ae9d-fb34f9aac449"))
+
+  private val facility = PatientMocker.facility(uuid = UUID.fromString("2a70f82e-92c6-4fce-b60e-6f083a8e725b"))
+  private val userSubject = PublishSubject.create<User>()
+
+  private val controller = BloodPressureEntrySheetController(
+      bloodPressureRepository = bloodPressureRepository,
+      appointmentRepository = appointmentRepository,
+      patientRepository = patientRepository,
+      dateValidator = dateValidator,
+      bpValidator = bpValidator,
+      userClock = testUserClock,
+      inputDatePaddingCharacter = UserInputDatePaddingCharacter('0'),
+      userSession = userSession,
+      facilityRepository = facilityRepository)
+
+  @Before
+  fun setUp() {
+    RxJavaPlugins.setIoSchedulerHandler { Schedulers.trampoline() }
+
+    whenever(dateValidator.dateInUserTimeZone()).doReturn(LocalDate.now(testUtcClock))
+    whenever(userSession.requireLoggedInUser()).doReturn(userSubject)
+    whenever(facilityRepository.currentFacility(user)).doReturn(Observable.just(facility))
+
+    uiEvents
+        .compose(controller)
+        .subscribe { uiChange -> uiChange(ui) }
+
+    userSubject.onNext(user)
+  }
+
+  // TODO(rj): 2019-10-10 There isn't a straight-forward way to set a time in the date validator as of now. We'll rely on mocks to test this code path (showDateIsInFutureError) for now.
+  @Test
+  @Suppress("IMPLICIT_CAST_TO_ANY")
+  @Parameters(method = "params for showing date validation errors")
+  fun `when save is clicked, date entry is active and input is invalid then validation errors should be shown`(
+      testParams: InvalidDateTestParams
+  ) {
+    val (openAs, day, month, year, errorResult, uiChangeVerification) = testParams
+
+    whenever(dateValidator.validate(any(), any())).doReturn(errorResult)
+    whenever(bloodPressureRepository.measurement(any())).doReturn(Observable.never())
+
+    uiEvents.run {
+      onNext(SheetCreated(openAs = openAs))
+      onNext(ScreenChanged(BloodPressureEntrySheet.ScreenType.DATE_ENTRY))
+      onNext(DayChanged(day))
+      onNext(MonthChanged(month))
+      onNext(YearChanged(year.takeLast(2)))
+      onNext(SaveClicked)
+    }
+
+    verify(bloodPressureRepository, never()).saveMeasurement(any(), any(), any(), any(), any(), any())
+    verify(bloodPressureRepository, never()).updateMeasurement(any())
+    verify(ui, never()).setBpSavedResultAndFinish()
+
+    uiChangeVerification(ui)
```

Comment:
We can remove the `uiChangeVerification` from the test params since both of them just verify that the `showDateIsInFutureError` is called.

Reply:
I am not comfortable refactoring the tests more than we need to. Because, I have to ensure the test fails before I refactor the test. Then have to make it pass again. This is risky and time consuming. Plus, there are tests in `BloodPressureEntrySheetControllerTest` that use the same data class.